### PR TITLE
🐛 Enhance Serialization and Deserialization of Maps and Sets

### PR DESCRIPTION
**Pull request by AdelUnito**
### TL;DR
This pull request enhances the serialization of complex data structures like javascript Maps & Sets in the RedisCache class to correctly handle these structures, preventing their data loss during serialization and deserialization.

### Changes Made:

serializeComplexDataStructures Function:

- Improved the code to correctly serialize Maps and Sets, ensuring their data is preserved during storage.

deserializeComplexDataStructures Function:

- Enhanced the code to accurately recognize and deserialize Maps and Sets, preventing data loss during retrieval.

These changes ensure that when Maps and Sets are used within the RedisCache, their data is properly maintained, improving the overall reliability of the code.

### Testing:

The updated functions have been tested using a comprehensive suite of unit tests to confirm that they correctly handle Maps and Sets, complex objects, and primitive values.

### CODE REVIEW

File: src/lib/RedisCache.ts [hunk 1]
```diff
@@ -160,13 +162,33 @@ export class RedisCache extends CacheInstance {
     }
 
     if (value instanceof Object) {
-      return RedisCache.JSON_PREFIX + JSON.stringify(value);
+      const simplifiedObject = RedisCache.serializeComplexDataStructures(value);
+      return RedisCache.JSON_PREFIX + JSON.stringify(simplifiedObject);
```
Comment (christianblais) [inline src/lib/RedisCache.ts:166]: I kinda liked the simplicity of "just" calling `JSON.stringify` and call it a day. Now, I understand Set doesn't support JSONification, nor does Map, but that makes wonder if it wouldn't be a case for replacer and reviver functions. I'm thinking out loud here, but what would you think if instead of having this `serializeComplexDataStructures` function, we were to use something like;

```
return RedisCache.JSON_PREFIX + JSON.stringify(value, (key, value) => {
  if (value instanceof Set) {
    return { __type: 'Set', value: [...value] };
  } else if (value instanceof Map) {
    return { __type: 'Map', value: [...value] };
  } else {
    return value;
  }
});
```

To read the data back, maybe something like this?

```
return JSON.parse(value.substring(RedisCache.JSON_PREFIX.length), (key, value) => {
  if (typeof value === 'object' && value !== null) {
    if (value.__type === 'Set') {
      return new Set(value.value);
    } else if (value.__type === 'Map') {
      return new Map(value.value);
    } else {
      return value;
    }
  } else {
    return value;
  }
});
```

This would have the added benefit of being recursive by design. This wasn't tested, but it's just to give an idea.

**Review by erykwarren — COMMENTED** [review]
Have you considered using `JSON.stringify` and `parse` functionality of passing a replacer and a reviver?
https://stackoverflow.com/questions/29085197/how-do-you-json-stringify-an-es6-map

**Comment by AdrienDesola** [comment]
Wow, the timing of you two is excellent!

File: src/lib/RedisCache.ts (hunk 1 above)
Comment (AdrienDesola) [reply src/lib/RedisCache.ts:166]: >  This wasn't tested, but it's just to give an idea.

Tested it! And it works well :obamaimpressed:!

Maybe just don't use `[...value]` but `Array.from(` even though it [seems to only matter](https://medium.com/@ra100x/node-js-performance-array-from-set-vs-spread-set-1387080cd266) for array smaller than 10,  I could see it be the majority of our use cases and who doesn't want a 15% increase in pref for free anyway?


Good night gentleman

**Review by AdrienDesola — COMMENTED** [review]
.

**Comment by AdelUnito** [comment]
Thanks all for the late reviews. 
Night owls

File: src/lib/RedisCache.ts [hunk 2]
```diff
@@ -160,7 +162,15 @@ export class RedisCache extends CacheInstance {
     }
 
     if (value instanceof Object) {
-      return RedisCache.JSON_PREFIX + JSON.stringify(value);
+      return RedisCache.JSON_PREFIX + JSON.stringify(value,  (key, value) => {
+        if (value instanceof Set) {
+          return { __dataType: RedisCache.SET_PREFIX, value: Array.from(value) };
```
Comment (AdrienDesola) [inline src/lib/RedisCache.ts:167]: 🤔 not sure you need the MAP_PREFIX, SET_PREFIX here. It's already in a separate key __dataType

`__dataType: Set`, `__dataType: Get` should be enough imo

**Review by christianblais — APPROVED** [review]

**Review by ronjouch — APPROVED** [review]
LeuGeuTeuMeu

File: src/lib/RedisCache.ts (hunk 2 above)
Comment (christianblais) [reply src/lib/RedisCache.ts:167]: I was thinking along the same lines. Not a blocker for me, but using these prefixes confused me more than it should have.

**Review by AdrienDesola — CHANGES_REQUESTED** [review]
supa small nit 🙏

File: src/lib/RedisCache.ts [hunk 3]
```diff
@@ -205,7 +215,19 @@ export class RedisCache extends CacheInstance {
     }
 
     if (value.startsWith(RedisCache.JSON_PREFIX)) {
-      return JSON.parse(value.substring(RedisCache.JSON_PREFIX.length));
+      return JSON.parse(value.substring(RedisCache.JSON_PREFIX.length), (key, value) => {
+        if (typeof value === 'object' && value !== null) {
+          if (value.__dataType === RedisCache.SET_PREFIX) {
```
[the remark this answers is not in the document]
Comment (AdelUnito) [reply src/lib/RedisCache.ts]: I changed it, no strong opinion about that. Initially, I though it could be helpful to have the long UUI in case we fetch directly from Redis for debugging

File: src/lib/RedisCache.ts (hunk 2 above)
Comment (AdelUnito) [reply src/lib/RedisCache.ts]: https://github.com/unitoio/cachette/pull/77#discussion_r1378867068

**Review by AdrienDesola — APPROVED** [review]
noice! LGTM!
